client/allocrunner/groupservice_hook.go

```diff
@@ -3,30 +3,65 @@ package allocrunner
 import (
 	"sync"
 
-	hclog "github.com/hashicorp/go-hclog"
 	log "github.com/hashicorp/go-hclog"
 	"github.com/hashicorp/nomad/client/allocrunner/interfaces"
 	"github.com/hashicorp/nomad/client/consul"
+	"github.com/hashicorp/nomad/client/taskenv"
+	agentconsul "github.com/hashicorp/nomad/command/agent/consul"
 	"github.com/hashicorp/nomad/nomad/structs"
+	"github.com/hashicorp/nomad/plugins/drivers"
 )
 
 // groupServiceHook manages task group Consul service registration and
 // deregistration.
 type groupServiceHook struct {
-	alloc        *structs.Allocation
+	allocID      string
+	group        string
+	restarter    agentconsul.WorkloadRestarter
 	consulClient consul.ConsulServiceAPI
 	prerun       bool
-	mu           sync.Mutex
 
 	logger log.Logger
+
+	// The following fields may be updated
+	canary   bool
+	services []*structs.Service
+	networks structs.Networks
+	taskEnv  *taskenv.TaskEnv
+
+	// Since Update() may be called concurrently with any other hook all
+	// hook methods must be fully serialized
+	mu sync.Mutex
+}
+
+type groupServiceHookConfig struct {
+	alloc     *structs.Allocation
+	consul    consul.ConsulServiceAPI
+	restarter agentconsul.WorkloadRestarter
+	taskEnv   *taskenv.TaskEnv
+	logger    log.Logger
 }
 
-func newGroupServiceHook(logger hclog.Logger, alloc *structs.Allocation, consulClient consul.ConsulServiceAPI) *groupServiceHook {
+func newGroupServiceHook(cfg groupServiceHookConfig) *groupServiceHook {
 	h := &groupServiceHook{
-		alloc:        alloc,
-		consulClient: consulClient,
+		allocID:      cfg.alloc.ID,
+		group:        cfg.alloc.TaskGroup,
+		restarter:    cfg.restarter,
+		consulClient: cfg.consul,
+		taskEnv:      cfg.taskEnv,
+	}
+	h.logger = cfg.logger.Named(h.Name())
+	h.services = cfg.alloc.Job.LookupTaskGroup(h.group).Services
+
+	var networks structs.Networks
+	if cfg.alloc.AllocatedResources != nil {
+		networks = cfg.alloc.AllocatedResources.Shared.Networks
```

Comment:
```suggestion
		h.networks = cfg.alloc.AllocatedResources.Shared.Networks
```

I don't think you need the local networks var since its zero value is what h.networks is initialized to anyway.